authority/authority.go

```diff
@@ -600,20 +603,86 @@ func (a *Authority) init() error {
 			return admin.WrapErrorISE(err, "error loading provisioners to initialize authority")
 		}
 		if len(provs) == 0 && !strings.EqualFold(a.config.AuthorityConfig.DeploymentType, "linked") {
-			// Create First Provisioner
-			prov, err := CreateFirstProvisioner(ctx, a.adminDB, string(a.password))
-			if err != nil {
-				return admin.WrapErrorISE(err, "error creating first provisioner")
+			// Migration will currently only be kicked off once, because either one or more provisioners
+			// are migrated or a default JWK provisioner will be created in the DB. It won't run for
+			// linked or hosted deployments. Not for linked, because that case is explicitly checked
+			// for above. Not for hosted, because there'll be at least an existing OIDC provisioner.
+			var firstJWKProvisioner *linkedca.Provisioner
+			if len(a.config.AuthorityConfig.Provisioners) > 0 {
+				// Existing provisioners detected; try migrating them to DB storage.
+				a.initLogf("Starting migration of provisioners")
+				for _, p := range a.config.AuthorityConfig.Provisioners {
+					lp, err := ProvisionerToLinkedca(p)
+					if err != nil {
+						return admin.WrapErrorISE(err, "error transforming provisioner %q while migrating", p.GetName())
+					}
+
+					// Store the provisioner to be migrated
+					if err := a.adminDB.CreateProvisioner(ctx, lp); err != nil {
+						return admin.WrapErrorISE(err, "error creating provisioner %q while migrating", p.GetName())
+					}
+
+					// Mark the first JWK provisioner, so that it can be used for administration purposes
+					if firstJWKProvisioner == nil && lp.Type == linkedca.Provisioner_JWK {
+						firstJWKProvisioner = lp
+						a.initLogf("Migrated JWK provisioner %q with admin permissions", p.GetName())
+					} else {
+						a.initLogf("Migrated %s provisioner %q", p.GetType(), p.GetName())
+					}
+				}
+
+				c := a.config
+				if c.WasLoadedFromFile() {
+					// TODO(hs): check if prerequisites for writing files look OK (user/group, permission bits, etc) as
+					// extra safety check before trying to write at all?
+
+					// Remove the existing provisioners from the authority configuration
+					// and commit it to the existing configuration file. NOTE: committing
+					// the configuration at this point also writes other properties that
+					// have been initialized with default values, such as the `backdate` and
+					// `template` settings in the `authority`.
+					oldProvisioners := c.AuthorityConfig.Provisioners
+					c.AuthorityConfig.Provisioners = []provisioner.Interface{}
+					if err := c.Commit(); err != nil {
+						// Restore the provisioners in in-memory representation for consistency
+						// when writing the updated configuration fails. This is considered a soft
+						// error, so execution can continue.
+						c.AuthorityConfig.Provisioners = oldProvisioners
+						a.initLogf("Failed removing provisioners from configuration: %v", err)
+					}
+				}
```

Comment:
I don't think we should rewrite this file. The format will change, it might have some annotations, and it might be the source of truth for migrating the same initial configuration to a different server. We won't use it again, that for sure, but as long as we make this clear, docs, using logs, ...

Reply:
I'm not entirely with you on this one. Sure, the file will change, the format may be different, etc. But this is not the only place in which we programmatically update the contents. It'll also not be consistent. But I'll remove it.

Reply:
As far as I know, the other place is when you do `step ca provisioner add --ca-config ...`, and as this is executed by a user, so he can expect that file to change. This code changes the file without any user interaction.